File: cpp/src/agent_context.cpp
#include "agent_context.hpp"
#include "convert.hpp"

namespace longbridge {
namespace agent {

using longbridge::convert::convert;

namespace {

/// Flattened storage for an `AnswersByToolCall`, matching the C layer's
/// array-of-entries shape (`lb_answers_by_tool_call_entry_t` /
/// `lb_answer_question_t`). The `entries`/`per_entry_answers` vectors are
/// reserved up front so pointers into already-pushed elements stay valid
/// while more are appended; all `const char*`s borrow from the original
/// `AnswersByToolCall`, so this must not outlive it.
struct AnswersFFI
{
  std::vector<lb_answers_by_tool_call_entry_t> entries;
  std::vector<std::vector<lb_answer_question_t>> per_entry_answers;
};

AnswersFFI
build_answers_ffi(const AnswersByToolCall& answers)
{
  AnswersFFI ffi;
  ffi.entries.reserve(answers.size());
  ffi.per_entry_answers.reserve(answers.size());
  for (const auto& [tool_call_id, qa] : answers) {
    std::vector<lb_answer_question_t> qs;
    qs.reserve(qa.size());
    for (const auto& [question, answer] : qa) {
      qs.push_back(lb_answer_question_t{ question.c_str(), answer.c_str() });
    }
    ffi.per_entry_answers.push_back(std::move(qs));
    const auto& stored = ffi.per_entry_answers.back();
    ffi.entries.push_back(lb_answers_by_tool_call_entry_t{
      tool_call_id.c_str(), stored.data(), stored.size() });
  }
  return ffi;
}

} // namespace

AgentContext::AgentContext()
  : ctx_(nullptr)
{
}

AgentContext::AgentContext(const lb_agent_context_t* ctx)
{
  ctx_ = ctx;
  if (ctx_) {
    lb_agent_context_retain(ctx_);
  }
}

AgentContext::AgentContext(const AgentContext& ctx)
{
  ctx_ = ctx.ctx_;
  if (ctx_) {
    lb_agent_context_retain(ctx_);
  }
}

AgentContext::AgentContext(AgentContext&& ctx)
{
  ctx_ = ctx.ctx_;
  ctx.ctx_ = nullptr;
}

AgentContext::~AgentContext()
{
  if (ctx_) {
    lb_agent_context_release(ctx_);
  }
}

AgentContext&
AgentContext::operator=(const AgentContext& ctx)
{
  ctx_ = ctx.ctx_;
  if (ctx_) {
    lb_agent_context_retain(ctx_);
  }
  return *this;
}

AgentContext
AgentContext::create(const Config& config)
{
  auto* ctx_ptr = lb_agent_context_new(config);
  AgentContext ctx(ctx_ptr);
  if (ctx_ptr) {
    lb_agent_context_release(ctx_ptr);
  }
  return ctx;
}

void
AgentContext::workspaces(
  AsyncCallback<AgentContext, WorkspacesResponse> callback) const
{
  lb_agent_context_workspaces(
    ctx_,
    [](auto res) {
      auto callback_ptr =
        callback::get_async_callback<AgentContext, WorkspacesResponse>(
          res->userdata);
      AgentContext ctx((const lb_agent_context_t*)res->ctx);
      Status status(res->error);

      if (status) {
        WorkspacesResponse resp =
          convert((const lb_workspaces_response_t*)res->data);
        (*callback_ptr)(AsyncResult<AgentContext, WorkspacesResponse>(
          ctx, std::move(status), &resp));
      } else {
        (*callback_ptr)(AsyncResult<AgentContext, WorkspacesResponse>(
          ctx, std::move(status), nullptr));
      }
    },
    new AsyncCallback<AgentContext, WorkspacesResponse>(callback));
}

void
AgentContext::agents(
  const std::string& workspace_id,
  const std::optional<GetAgentsOptions>& opts,
  AsyncCallback<AgentContext, AgentsResponse> callback) const
{
  lb_get_agents_options_t opts2 = { nullptr, nullptr, nullptr };
  if (opts) {
    opts2.page = opts->page ? &opts->page.value() : nullptr;
    opts2.limit = opts->limit ? &opts->limit.value() : nullptr;
    opts2.name = opts->name ? opts->name->c_str() : nullptr;
  }

  lb_agent_context_agents(
    ctx_,
    workspace_id.c_str(),
    &opts2,
    [](auto res) {
      auto callback_ptr =
        callback::get_async_callback<AgentContext, AgentsResponse>(
          res->userdata);
      AgentContext ctx((const lb_agent_context_t*)res->ctx);
      Status status(res->error);

      if (status) {
        AgentsResponse resp = convert((const lb_agents_response_t*)res->data);
        (*callback_ptr)(AsyncResult<AgentContext, AgentsResponse>(
          ctx, std::move(status), &resp));
      } else {
        (*callback_ptr)(AsyncResult<AgentContext, AgentsResponse>(
          ctx, std::move(status), nullptr));
      }
    },
    new AsyncCallback<AgentContext, AgentsResponse>(callback));
}

void
AgentContext::conversation(
  const std::string& agent_id,
  const std::string& query,
  const std::optional<std::string>& chat_uid,
  AsyncCallback<AgentContext, ConversationResponse> callback) const
{
  lb_agent_context_conversation(
    ctx_,
    agent_id.c_str(),
    query.c_str(),
    chat_uid ? chat_uid->c_str() : nullptr,
    [](auto res) {
      auto callback_ptr =
        callback::get_async_callback<AgentContext, ConversationResponse>(
          res->userdata);
      AgentContext ctx((const lb_agent_context_t*)res->ctx);
      Status status(res->error);

      if (status) {
        ConversationResponse resp =
          convert((const lb_conversation_response_t*)res->data);
        (*callback_ptr)(AsyncResult<AgentContext, ConversationResponse>(
          ctx, std::move(status), &resp));
      } else {
        (*callback_ptr)(AsyncResult<AgentContext, ConversationResponse>(
          ctx, std::move(status), nullptr));
      }
    },
    new AsyncCallback<AgentContext, ConversationResponse>(callback));
}

void
AgentContext::continue_conversation(
  const std::string& agent_id,
  const std::string& chat_uid,
  const std::string& message_id,
  const AnswersByToolCall& answers,
  AsyncCallback<AgentContext, ConversationResponse> callback) const
{
  AnswersFFI ffi = build_answers_ffi(answers);

  lb_agent_context_continue_conversation(
    ctx_,
    agent_id.c_str(),
    chat_uid.c_str(),
    message_id.c_str(),
    ffi.entries.empty() ? nullptr : ffi.entries.data(),
    ffi.entries.size(),
    [](auto res) {
      auto callback_ptr =
        callback::get_async_callback<AgentContext, ConversationResponse>(
          res->userdata);
      AgentContext ctx((const lb_agent_context_t*)res->ctx);
      Status status(res->error);

      if (status) {
        ConversationResponse resp =
          convert((const lb_conversation_response_t*)res->data);
        (*callback_ptr)(AsyncResult<AgentContext, ConversationResponse>(
          ctx, std::move(status), &resp));
      } else {
        (*callback_ptr)(AsyncResult<AgentContext, ConversationResponse>(
          ctx, std::move(status), nullptr));
      }
    },
    new AsyncCallback<AgentContext, ConversationResponse>(callback));
}

void
AgentContext::conversation_streamed(
  const std::string& agent_id,
  const std::string& query,
  const std::optional<std::string>& chat_uid,
  PushCallback<AgentContext, ConversationStreamEvent> on_event,
  AsyncCallback<AgentContext, ConversationResponse> callback) const
{
  lb_agent_context_conversation_streamed(
    ctx_,
    agent_id.c_str(),
    query.c_str(),
    chat_uid ? chat_uid->c_str() : nullptr,
    [](auto ctx, auto event, auto userdata) {
      auto cb =
        callback::get_push_callback<AgentContext, ConversationStreamEvent>(
          userdata);
      ConversationStreamEvent event2 = convert(event);
      (*cb)(PushEvent<AgentContext, ConversationStreamEvent>(
        AgentContext(ctx), &event2));
    },
    new PushCallback<AgentContext, ConversationStreamEvent>(on_event),
    [](auto p) {
      delete (PushCallback<AgentContext, ConversationStreamEvent>*)p;
    },
    [](auto res) {
      auto callback_ptr =
        callback::get_async_callback<AgentContext, ConversationResponse>(
          res->userdata);
      AgentContext ctx((const lb_agent_context_t*)res->ctx);
      Status status(res->error);

      if (status) {
        ConversationResponse resp =
          convert((const lb_conversation_response_t*)res->data);
        (*callback_ptr)(AsyncResult<AgentContext, ConversationResponse>(
          ctx, std::move(status), &resp));
      } else {
        (*callback_ptr)(AsyncResult<AgentContext, ConversationResponse>(
          ctx, std::move(status), nullptr));
      }
    },
    new AsyncCallback<AgentContext, ConversationResponse>(callback));
}

void
AgentContext::continue_conversation_streamed(
  const std::string& agent_id,
  const std::string& chat_uid,
  const std::string& message_id,
  const AnswersByToolCall& answers,
  PushCallback<AgentContext, ConversationStreamEvent> on_event,
  AsyncCallback<AgentContext, ConversationResponse> callback) const
{
  AnswersFFI ffi = build_answers_ffi(answers);

  lb_agent_context_continue_conversation_streamed(
    ctx_,
    agent_id.c_str(),
    chat_uid.c_str(),
    message_id.c_str(),
    ffi.entries.empty() ? nullptr : ffi.entries.data(),
    ffi.entries.size(),
    [](auto ctx, auto event, auto userdata) {
      auto cb =
        callback::get_push_callback<AgentContext, ConversationStreamEvent>(
          userdata);
      ConversationStreamEvent event2 = convert(event);
      (*cb)(PushEvent<AgentContext, ConversationStreamEvent>(
        AgentContext(ctx), &event2));
    },
    new PushCallback<AgentContext, ConversationStreamEvent>(on_event),
    [](auto p) {
      delete (PushCallback<AgentContext, ConversationStreamEvent>*)p;
    },
    [](auto res) {
      auto callback_ptr =
        callback::get_async_callback<AgentContext, ConversationResponse>(
          res->userdata);
      AgentContext ctx((const lb_agent_context_t*)res->ctx);
      Status status(res->error);

      if (status) {
        ConversationResponse resp =
          convert((const lb_conversation_response_t*)res->data);
        (*callback_ptr)(AsyncResult<AgentContext, ConversationResponse>(
          ctx, std::move(status), &resp));
      } else {
        (*callback_ptr)(AsyncResult<AgentContext, ConversationResponse>(
          ctx, std::move(status), nullptr));
      }
    },
    new AsyncCallback<AgentContext, ConversationResponse>(callback));
}

} // namespace agent
} // namespace longbridge

File: java/src/agent_context.rs
use std::sync::Arc;

use jni::{
    JNIEnv, JavaVM,
    objects::{GlobalRef, JClass, JObject, JString, JValue},
};
use longbridge::{
    AgentContext, Config,
    agent::{self, AnswersByToolCall, ConversationStreamEvent, GetAgentsOptions},
};

use crate::{
    async_util,
    error::{JniError, jni_result},
    init::CONVERSATION_STREAM_SUBSCRIPTION_CLASS,
    types::{FromJValue, IntoJValue, JavaInteger, get_field},
};

struct ContextObj {
    ctx: AgentContext,
}

/// Everything needed to deliver `Flow.Subscriber` callbacks (`onNext` /
/// `onError` / `onComplete`) from a background tokio task back into the JVM.
/// Shared (via `Arc`) between the three closures handed to
/// [`agent::ConversationStreamSubscription::spawn`], since they all target
/// the same `Subscriber` instance.
struct SubscriberSink {
    jvm: JavaVM,
    subscriber: GlobalRef,
}

fn deliver_on_next(sink: &SubscriberSink, event: ConversationStreamEvent) {
    let Ok(mut env) = sink.jvm.attach_current_thread() else {
        return;
    };
    if let Ok(value) = event.into_jvalue(&mut env) {
        // `Flow.Subscriber<T>.onNext(T)` is generic; the JVM always exposes a
        // bridge method with the erased `(Ljava/lang/Object;)V` descriptor
        // for any concrete implementation, so this is the correct signature
        // to target regardless of what `T` the caller's `Subscriber` names.
        let _ = env.call_method(
            &sink.subscriber,
            "onNext",
            "(Ljava/lang/Object;)V",
            &[value.borrow()],
        );
    }
}

fn deliver_on_error(sink: &SubscriberSink, err: longbridge::Error) {
    let Ok(mut env) = sink.jvm.attach_current_thread() else {
        return;
    };
    let err_obj = JniError::from(err).into_error_object(&mut env);
    let _ = env.call_method(
        &sink.subscriber,
        "onError",
        "(Ljava/lang/Throwable;)V",
        &[JValue::from(&err_obj)],
    );
}

fn deliver_on_complete(sink: &SubscriberSink) {
    let Ok(mut env) = sink.jvm.attach_current_thread() else {
        return;
    };
    let _ = env.call_method(&sink.subscriber, "onComplete", "()V", &[]);
}

fn new_subscription_object<'a>(
    env: &mut JNIEnv<'a>,
    handle: i64,
) -> jni::errors::Result<JObject<'a>> {
    let cls = CONVERSATION_STREAM_SUBSCRIPTION_CLASS.get().unwrap();
    env.new_object(cls, "(J)V", &[JValue::from(handle)])
}

/// Finish a successful `subscribe()`: box up `subscription` behind an opaque
/// handle, wrap it in a Java `ConversationStreamSubscription`, and deliver it
/// via `onSubscribe`. If attaching to the JVM or constructing the Java object
/// fails, the boxed subscription is freed immediately so it isn't leaked (the
/// subscriber will simply never hear back, which is the best we can do at
/// that point).
fn finish_subscribe_success(
    sink: Arc<SubscriberSink>,
    subscription: agent::ConversationStreamSubscription,
) {
    let handle = Box::into_raw(Box::new(subscription)) as i64;

    let Ok(mut env) = sink.jvm.attach_current_thread() else {
        unsafe {
            let _ = Box::from_raw(handle as *mut agent::ConversationStreamSubscription);
        }
        return;
    };

    match new_subscription_object(&mut env, handle) {
        Ok(obj) => {
            let _ = env.call_method(
                &sink.subscriber,
                "onSubscribe",
                "(Ljava/util/concurrent/Flow$Subscription;)V",
                &[JValue::from(&obj)],
            );
        }
        Err(_) => unsafe {
            let _ = Box::from_raw(handle as *mut agent::ConversationStreamSubscription);
        },
    }
}

fn read_get_agents_options(
    env: &mut JNIEnv,
    opts: &JObject,
) -> jni::errors::Result<Option<GetAgentsOptions>> {
    if opts.is_null() {
        return Ok(None);
    }

    let mut new_opts = GetAgentsOptions::new();
    let page: Option<JavaInteger> = get_field(env, opts, "page")?;
    if let Some(page) = page {
        new_opts = new_opts.page(page.into());
    }
    let limit: Option<JavaInteger> = get_field(env, opts, "limit")?;
    if let Some(limit) = limit {
        new_opts = new_opts.limit(limit.into());
    }
    let name: Option<String> = get_field(env, opts, "name")?;
    if let Some(name) = name {
        new_opts = new_opts.name(name);
    }
    Ok(Some(new_opts))
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_newAgentContext(
    mut env: JNIEnv,
    _class: JClass,
    config: i64,
) -> i64 {
    jni_result(&mut env, 0i64, |_env| {
        let config = Arc::new((*(config as *const Config)).clone());
        Ok(Box::into_raw(Box::new(ContextObj {
            ctx: AgentContext::new(config),
        })) as i64)
    })
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_freeAgentContext(
    _env: JNIEnv,
    _class: JClass,
    ctx: i64,
) {
    let _ = Box::from_raw(ctx as *mut ContextObj);
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_agentContextWorkspaces(
    mut env: JNIEnv,
    _class: JClass,
    context: i64,
    callback: JObject,
) {
    jni_result(&mut env, (), |env| {
        let context = &*(context as *const ContextObj);
        async_util::execute(
            env,
            callback,
            async move { Ok(context.ctx.workspaces().await?) },
        )?;
        Ok(())
    })
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_agentContextAgents(
    mut env: JNIEnv,
    _class: JClass,
    context: i64,
    workspace_id: JString,
    opts: JObject,
    callback: JObject,
) {
    jni_result(&mut env, (), |env| {
        let context = &*(context as *const ContextObj);
        let workspace_id: String = FromJValue::from_jvalue(env, workspace_id.into())?;
        let opts = read_get_agents_options(env, &opts)?;
        async_util::execute(env, callback, async move {
            Ok(context.ctx.agents(workspace_id, opts).await?)
        })?;
        Ok(())
    })
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_agentContextConversation(
    mut env: JNIEnv,
    _class: JClass,
    context: i64,
    agent_id: JString,
    query: JString,
    chat_uid: JObject,
    callback: JObject,
) {
    jni_result(&mut env, (), |env| {
        let context = &*(context as *const ContextObj);
        let agent_id: String = FromJValue::from_jvalue(env, agent_id.into())?;
        let query: String = FromJValue::from_jvalue(env, query.into())?;
        let chat_uid: Option<String> = FromJValue::from_jvalue(env, chat_uid.into())?;
        async_util::execute(env, callback, async move {
            Ok(crate::types::ConversationResponse::from(
                context.ctx.conversation(agent_id, query, chat_uid).await?,
            ))
        })?;
        Ok(())
    })
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_agentContextContinueConversation(
    mut env: JNIEnv,
    _class: JClass,
    context: i64,
    agent_id: JString,
    chat_uid: JString,
    message_id: JString,
    answers_json: JString,
    callback: JObject,
) {
    jni_result(&mut env, (), |env| {
        let context = &*(context as *const ContextObj);
        let agent_id: String = FromJValue::from_jvalue(env, agent_id.into())?;
        let chat_uid: String = FromJValue::from_jvalue(env, chat_uid.into())?;
        let message_id: String = FromJValue::from_jvalue(env, message_id.into())?;
        let answers_json: String = FromJValue::from_jvalue(env, answers_json.into())?;
        let answers: AnswersByToolCall = serde_json::from_str(&answers_json).unwrap_or_default();
        async_util::execute(env, callback, async move {
            Ok(crate::types::ConversationResponse::from(
                context
                    .ctx
                    .continue_conversation(agent_id, chat_uid, message_id, answers)
                    .await?,
            ))
        })?;
        Ok(())
    })
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_agentContextConversationStreamSubscribe(
    mut env: JNIEnv,
    _class: JClass,
    context: i64,
    agent_id: JString,
    query: JString,
    chat_uid: JObject,
    subscriber: JObject,
) {
    jni_result(&mut env, (), |env| {
        let context = &*(context as *const ContextObj);
        let ctx = context.ctx.clone();
        let agent_id: String = FromJValue::from_jvalue(env, agent_id.into())?;
        let query: String = FromJValue::from_jvalue(env, query.into())?;
        let chat_uid: Option<String> = FromJValue::from_jvalue(env, chat_uid.into())?;
        let jvm = env.get_java_vm()?;
        let subscriber = env.new_global_ref(subscriber)?;

        // `subscribe()` must return immediately without doing I/O (cold
        // Publisher semantics) — the actual connect happens on the shared
        // runtime, and `onSubscribe`/`onError` are delivered asynchronously
        // once it resolves.
        longbridge::runtime_handle().spawn(async move {
            let sink = Arc::new(SubscriberSink { jvm, subscriber });
            match ctx.conversation_streamed(agent_id, query, chat_uid).await {
                Ok(stream) => {
                    let (sink_next, sink_err, sink_complete) =
                        (sink.clone(), sink.clone(), sink.clone());
                    let subscription = agent::ConversationStreamSubscription::spawn(
                        stream,
                        move |event| deliver_on_next(&sink_next, event),
                        move |err| deliver_on_error(&sink_err, err),
                        move || deliver_on_complete(&sink_complete),
                    );
                    finish_subscribe_success(sink, subscription);
                }
                Err(err) => deliver_on_error(&sink, err),
            }
        });
        Ok(())
    })
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_agentContextContinueConversationStreamSubscribe(
    mut env: JNIEnv,
    _class: JClass,
    context: i64,
    agent_id: JString,
    chat_uid: JString,
    message_id: JString,
    answers_json: JString,
    subscriber: JObject,
) {
    jni_result(&mut env, (), |env| {
        let context = &*(context as *const ContextObj);
        let ctx = context.ctx.clone();
        let agent_id: String = FromJValue::from_jvalue(env, agent_id.into())?;
        let chat_uid: String = FromJValue::from_jvalue(env, chat_uid.into())?;
        let message_id: String = FromJValue::from_jvalue(env, message_id.into())?;
        let answers_json: String = FromJValue::from_jvalue(env, answers_json.into())?;
        let answers: AnswersByToolCall = serde_json::from_str(&answers_json).unwrap_or_default();
        let jvm = env.get_java_vm()?;
        let subscriber = env.new_global_ref(subscriber)?;

        longbridge::runtime_handle().spawn(async move {
            let sink = Arc::new(SubscriberSink { jvm, subscriber });
            match ctx
                .continue_conversation_streamed(agent_id, chat_uid, message_id, answers)
                .await
            {
                Ok(stream) => {
                    let (sink_next, sink_err, sink_complete) =
                        (sink.clone(), sink.clone(), sink.clone());
                    let subscription = agent::ConversationStreamSubscription::spawn(
                        stream,
                        move |event| deliver_on_next(&sink_next, event),
                        move |err| deliver_on_error(&sink_err, err),
                        move || deliver_on_complete(&sink_complete),
                    );
                    finish_subscribe_success(sink, subscription);
                }
                Err(err) => deliver_on_error(&sink, err),
            }
        });
        Ok(())
    })
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_conversationStreamSubscriptionRequest(
    _env: JNIEnv,
    _class: JClass,
    handle: i64,
    n: i64,
) {
    let subscription = &*(handle as *const agent::ConversationStreamSubscription);
    subscription.request(n.max(0) as u64);
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_conversationStreamSubscriptionCancel(
    _env: JNIEnv,
    _class: JClass,
    handle: i64,
) {
    let subscription = &*(handle as *const agent::ConversationStreamSubscription);
    subscription.cancel();
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_freeConversationStreamSubscription(
    _env: JNIEnv,
    _class: JClass,
    handle: i64,
) {
    let _ = Box::from_raw(handle as *mut agent::ConversationStreamSubscription);
}
